library_generation/model/config_change.py

```diff
@@ -0,0 +1,159 @@
+    GAPIC_ADDITION = 6
+    # As of Mar. 2024, we decide not to produce this type of change because we
+    # still need to manually remove the libray.
+    # GAPIC_REMOVAL = 7
+
+
+class HashLibrary:
+    """
+    Data class to group a LibraryConfig object and its hash value together.
+    """
+
+    def __init__(self, hash_value: int, library: LibraryConfig):
+        self.hash_value = hash_value
+        self.library = library
+
+
+class LibraryChange:
+    def __init__(self, changed_param: str, latest_value: str, library_name: str = ""):
+        self.changed_param = changed_param
+        self.latest_value = latest_value
+        self.library_name = library_name
+
+
+class QualifiedCommit:
+    def __init__(self, commit: Commit, libraries: set[str]):
+        self.commit = commit
+        self.libraries = libraries
+
+
+class ConfigChange:
+    ALL_LIBRARIES_CHANGED = None
+
+    def __init__(
+        self,
+        change_to_libraries: dict[ChangeType, list[LibraryChange]],
+        baseline_config: GenerationConfig,
+        latest_config: GenerationConfig,
+    ):
+        self.change_to_libraries = change_to_libraries
+        self.baseline_config = baseline_config
+        self.latest_config = latest_config
+
+    def get_changed_libraries(self) -> Optional[list[str]]:
+        """
+        Returns a unique, sorted list of library name of changed libraries.
+        None if there is a repository level change, which means all libraries
+        in the latest_config will be generated.
+        :return: library names of change libraries.
+        """
+        if ChangeType.REPO_LEVEL_CHANGE in self.change_to_libraries:
+            return ConfigChange.ALL_LIBRARIES_CHANGED
+        library_names = set()
+        for change_type, library_changes in self.change_to_libraries.items():
+            if change_type == ChangeType.GOOGLEAPIS_COMMIT:
+                library_names.update(self.__get_library_names_from_qualified_commits())
+            else:
+                library_names.update(
+                    [library_change.library_name for library_change in library_changes]
+                )
+        return sorted(list(library_names))
+
+    def get_qualified_commits(
+        self,
+        repo_url: str = "https://github.com/googleapis/googleapis.git",
+    ) -> list[QualifiedCommit]:
+        """
+        Returns qualified commits from configuration change.
+
+        A qualified commit is a commit that changes at least one file (excluding
+        BUILD.bazel) within a versioned proto path in the given proto_paths.
+        :param repo_url: the repository contains the commit history.
+        :return: QualifiedCommit objects.
+        """
+        tmp_dir = sh_util("get_output_folder")
+        shutil.rmtree(tmp_dir, ignore_errors=True)
+        os.mkdir(tmp_dir)
+        # we only need commit history, thus shadow clone is enough.
+        repo = Repo.clone_from(url=repo_url, to_path=tmp_dir, filter=["blob:none"])
```

Comment:
I know this was done in previous PRs. Based on our recent discussions, is it possible to get the commit history remotely? Without cloning the repo? Creating new temp folders could have other implications like preventing us from running images in parallel.

Reply:
I did some digging but I can't find one without `git clone`.

I don't think we need to focus on parallel execution as it's not the main goal of this PR. Also, we can achieve parallelism if we can randomize the temp folder so that no thread will clone the repo into the same folder.

Reply:
I don't think it's a focus either but something we should keep in mind. Creating temp folders and writing files has other implications, e.g. it's usually slower than making a request, it complicates the scripts that we have to clean the temp folders and files up etc. 
That being said, it's OK if there is no easy way at this moment. We can refactor how we pull the proto definitions in general in the future. e.g. I see that we are currently [downloading proto definitions for every library](https://github.com/googleapis/sdk-platform-java/blob/main/library_generation/utilities.py#L161-L194), where we could download the whole googleapis repo as the first step of generate_repo.py. This actually brings another question that if googleapis_commitish is overridden on library level, our release notes might be generated incorrectly. 
In short, this PR is good to go as it is, but we need to put the above tasks into consideration and track them somewhere.

Reply:
> I see that we are currently [downloading proto definitions for every library](https://github.com/googleapis/sdk-platform-java/blob/main/library_generation/utilities.py#L161-L194),

The googleapis repo is downloaded only once unless there are googleapis commit in library level.

Since this use case is rarely used, how about removing it?

Reply:
Based on discussion, we keep this piece of code as-is.